smart-queue/enqueue.js

```diff
@@ -0,0 +1,78 @@
+import * as QueueActionTypes from "./action-types";
+
+function validKey(key) {
+  if (key === null) {
+    console.log("Missing key, every queue action should have a key!");
+    return false;
+  }
+  return true;
+}
+
+function indexOfAction(array, key) {
+  return array.findIndex(
+    ({ meta: { offline: { queue = null } } }) => queue && queue.key === key
+  );
+}
+export function enqueue(array, action) {
+  const { meta: { offline: { queue = null } } } = action;
+  let index, existingAction;
+
+  if (!queue) {
+    return [...array, action];
+  }
+
+  const { method = "UNKNOWN", key = null } = queue;
+  if (!validKey(key)) {
+    return array;
+  }
+
+  index = indexOfAction(array, key);
+
+  switch (method) {
+    case QueueActionTypes.CREATE:
+      if (index !== -1) {
```

Comment:
If this is the case then isn't there something wrong? I thought each create action would have a unique key.

Reply:
Theoretically yes, every `CREATE` action will have a unique key, but by any chance, if user dispatch CREATE action twice with the same key, we will end up having a duplicate entry in the `outbox`, which will effect subsequence `UPDATE` and `DELETE` operation for the same key.

Reply:
Might want to provide some feedback in this case. To let the user know they are doing something wrong.